File: Core/MasterClock.cs
using System;
using System.IO;
using System.Threading;
using System.Diagnostics;

namespace SETUE.Core
{
    public static class MasterClock
    {
        private static float _tickInterval = 0.012f;
        private static bool _running = false;
        private static Thread? _thread;
        public static readonly object LockObject = new object();
        private static Stopwatch _stopwatch = new Stopwatch();

        public static double CurrentTime => _stopwatch.Elapsed.TotalSeconds;

        public static void Load()
        {
            string path = "Core/MasterClock.csv";
            if (File.Exists(path))
            {
                var lines = File.ReadAllLines(path);
                if (lines.Length > 1)
                {
                    var parts = lines[1].Split(',');
                    if (parts.Length >= 2)
                    {
                        _tickInterval = float.Parse(parts[1]);
                    }
                }
            }
            Console.WriteLine($"[MasterClock] Tick interval = {_tickInterval} seconds");
        }

        public static void Start()
        {
            if (_running) return;
            _running = true;
            _stopwatch.Start();
            _thread = new Thread(Run);
            _thread.Start();
            Console.WriteLine("[MasterClock] Started");
        }

        public static void Stop()
        {
            _running = false;
            _thread?.Join();
            _stopwatch.Stop();
            Console.WriteLine("[MasterClock] Stopped");
        }

        public static void WaitForExit()
        {
            _thread?.Join();
        }

        private static void Run()
        {
            // === BOOT PHASE (runs once on the scheduler thread) ===
            Schedulers.RunBoot();

            // === MAIN LOOP ===
            while (_running)
            {
                double tickStart = CurrentTime;

                lock (LockObject)
                {
                    Schedulers.Update(CurrentTime);
                }

                double elapsed = CurrentTime - tickStart;
                int sleepMs = (int)((_tickInterval - elapsed) * 1000);
                if (sleepMs > 0)
                    Thread.Sleep(sleepMs);
            }
        }
    }
}

File: Core/MasterClock.ts
// src/Core/MasterClock.ts
import { parseSettingsCsv } from '../parseCsv.js';
import { Scheduler } from './Scheduler.js';

export class MasterClock {
  private static tickInterval: number = 0;
  private static running: boolean = false;
  private static lastTickTime: number = 0;

  static async load(csvPath: string = 'Core/MasterClock.csv') {
    const response = await fetch(csvPath);
    if (!response.ok) throw new Error(`[MasterClock] Missing ${csvPath}`);
    const csvText = await response.text();
    const settings = parseSettingsCsv(csvText);

    if (settings['tick_interval'] === undefined)
      throw new Error('[MasterClock] tick_interval not found');
    this.tickInterval = parseFloat(settings['tick_interval'] as string);
    if (isNaN(this.tickInterval) || this.tickInterval <= 0)
      throw new Error(`[MasterClock] Invalid tick_interval`);

    console.log(`[MasterClock] Tick interval = ${this.tickInterval} seconds`);
    await Scheduler.load('Core/Scheduler.csv');
  }

  static async start() {                       // <-- now async
    if (this.running) return;
    if (this.tickInterval <= 0) throw new Error('[MasterClock] tickInterval not loaded');

    this.running = true;
    this.lastTickTime = performance.now() / 1000;

    console.log('[MasterClock] Started – running boot first');
    await Scheduler.runBoot();                 // <-- await the whole boot sequence
    this.scheduleNext();
  }

  static stop() {
    this.running = false;
    console.log('[MasterClock] Stopped');
  }

  private static scheduleNext() {
    requestAnimationFrame(() => {
      if (!this.running) return;

      const now = performance.now() / 1000;
      const delta = now - this.lastTickTime;

      if (delta >= this.tickInterval) {
        const steps = Math.min(Math.floor(delta / this.tickInterval), 5);
        for (let i = 0; i < steps; i++) {
          const tickTime = this.lastTickTime + this.tickInterval;
          Scheduler.update(tickTime);
          this.lastTickTime = tickTime;
        }
      }
      this.scheduleNext();
    });
  }
}
